src/java.base/share/classes/java/lang/reflect/code/bytecode/BytecodeGenerator.java

```diff
@@ -357,10 +361,60 @@ private static boolean canDefer(ConstantOp op) {
         return !op.resultType().equals(JavaType.J_L_CLASS);
     }
 
-    // Var with a single-use entry block parameter operand can be deferred
+    // Single-use var or var with a single-use entry block parameter operand can be deferred
     private static boolean canDefer(VarOp op) {
         return !moreThanOneUse(op.result())
-            || op.operands().getFirst() instanceof Block.Parameter bp && bp.declaringBlock().isEntryBlock() && !moreThanOneUse(bp);
+            || op.operands().getFirst() instanceof Block.Parameter bp && bp.declaringBlock().isEntryBlock() && !moreThanOneUse(bp)
+            || canDeferVarCombo(op);
+
+    }
+
+    // Detection of var declaration in a dominant block, initialized with a redundant default value
+    // Combo of ConstantOp and VarOp can be deferred if all its VarLoadOp are dominated by VarStoreOp
+    private static boolean canDeferVarCombo(VarOp op) {
+        if (op.initOperand() instanceof Op.Result or && or.op() instanceof ConstantOp cop && canDefer(cop)) {
+            Set<Op.Result> allUses = op.result().uses();
+            Set<Op.Result> storeResults = allUses.stream().filter(r -> r.op() instanceof VarAccessOp.VarStoreOp).collect(Collectors.toSet());
+            // All VarLoadOp must be dominated by a VarStoreOp
+            for (Op.Result loadResult : allUses) {
+                if (loadResult.op() instanceof VarAccessOp.VarLoadOp) {
+                    if (!isDominatedBy(loadResult, storeResults)) {
+                        return false;
+                    }
+                }
+            }
+            return true;
+        }
+        return false;
+    }
+
+    // @@@ Test for dominant set
+    private static boolean isDominatedBy(Op.Result n, Set<Op.Result> doms) {
+        for (Op.Result dom : doms) {
+            if (n.isDominatedBy(dom)) {
+                return true;
+            }
+        }
+
+        Set<Block> stopBlocks = new HashSet<>();
```

Comment:
If i got this right, we are dealing with an equivalent of an uninitialized variable, of a primitive type, in source that will assigned in divergent code paths, and the compiler will check it's DA. In the model we currently require the modelling var op be initialized with a default constant value for the primitive type. We don't currently have a concept of an uninitialized value (which might help?), so we have to detect that all loads are dominated by stores, some of which are initializing stores (which detects the uninitialized case or a redundant initialization).

This specific bit of code after the dominated by checks determines if there is a path from `n`'s declaring block to the entry block that does not pass through any of `dom`'s declaring blocks? Why is that is important?

Reply:
Yes, this is about uninitialized variables (of any type), where there are two or more assignments in divergent paths. Concept of uninitialized variable will definitely help to avoid obsolete default constants declaration. This is a detection of such situation, so the default initialization is skipped and the roundtrip is stable.

Reply:
This code detect there is no path from a VarLoadOp to the entry block that does not pass any of the VarStoreOps. The default initialization is necessary if there is such path.

Reply:
I still don't understand why the second check is needed, traversing the predecessors of `n`'s declaring block. If `n` is not dominated by any values in set `doms` then presumably such traversal has to encounter the entry block? 

The implementation of `Value::isDominatedBy` defers to block domination:
```
    public boolean isDominatedBy(Value dom) {
        if (this == dom) {
            return true;
        }

        if (declaringBlock() != dom.declaringBlock()) {
            return declaringBlock().isDominatedBy(dom.declaringBlock());
        }
...
```

Reply:
Individual dominations are handled as you describe. It applies for cases like this:
```
entry -> var -> var store -> var load
            \
              > var store -> var load 
```

However following case needs a test for dominant set:
```
entry -> var -> var store -> var load
            \             /
              > var store
```
The var load is not dominated by any individual var store, however it is dominated by them.

The test should fail for example on following case:
```
entry -> var -> var store -> var load
             \            /     
                ---------
```

The reason why the second part is focused on blocks only is because ops cannot form a bypass inside one block.

Reply:
Code with redundant default value assignment is valid, it just causes instability (the code grows with each round).